text/chapter4.md

```diff
@@ -314,9 +314,9 @@ The keyword `do` introduces a block of code which uses do notation. The block co
 - Expressions which do not bind elements of the array to names. The last line `pure [i, j]` is an example of this kind of expression.
 - Expressions which give names to expressions, using the `let` keyword.
 
-This new notation hopefully makes the structure of the algorithm clearer. If you mentally replace the arrow `<-` with the word "choose", you might read it as follows: "choose an element `i` between 1 and n, then choose an element `j` between `i` and `n`, and return `[i, j]`".
+This new notation hopefully makes the structure of the algorithm clearer. If you mentally replace the arrow `<-` with the word "choose", you might read it as follows: "choose an element `i` between 1 and n, then choose an element `j` between `i` and `n`, and return the desired results, `i` and `j`, as `[i, j]`.
 
-In the last line, we use the `pure` function. This function can be evaluated in PSCi, but we have to provide a type:
```

Comment:
I'm uneasy with this change too. What does "by default" mean here? It feels to me like it will encourage readers to always use `pure` at the end of a `do` block without thinking about it, which seems antithetical to the goals of this section (i.e. we should be encouraging them to understand what it is actually doing). Also, `do` blocks often end with something other than `pure`, and this is perfectly acceptable; we don't want people to think that ending a `do` block with something other than `pure` is un-idiomatic.

Reply:
Again, I'm relating with how I struggled with this line `pure [i, j]`, ; a lot of new concepts are being introduced right here:

1. The word `pure`.
1. Maybe use a singleton instead of `pure`.  Why?
1. An array to wrap `I` and `j`; why using an array?
1. Result for what? 

I found I was mired in these details without really knowing what the purpose of the last entry in the `do` is for.

So, whether we use

> In the last line, we use the `pure` function. This function can be evaluated in PSCi, but we have to provide a type:

or

> By default, we apply the `pure` function to the results. This function can be evaluated in PSCi, but we have to provide a type:

I'm already lost.

In struggling through this, I realize that I'd be happy to undo all of this if I can see the `do` return value explained earlier.  What would help me out is to make the change earlier as follows:  Replace this

---

The keyword `do` introduces a block of code which uses do notation. The block consists of expressions of a few types:

- Expressions which bind elements of an array to a name. These are indicated with the backwards-facing arrow `<-`, with a name on the left, and an expression on the right whose type is an array.
- Expressions which do not bind elements of the array to names. The last line `pure [i, j]` is an example of this kind of expression.
- Expressions which give names to expressions, using the `let` keyword.

---

with (addition in **bold**)

---

The keyword `do` introduces a block of code which uses do notation. The block consists of expressions of a few types:

- Expressions which bind elements of an array to a name. These are indicated with the backwards-facing arrow `<-`, with a name on the left, and an expression on the right whose type is an array.
- Expressions which do not bind elements of the array to names. **The `do` result is an example of this kind of expression and is illustrated in the last line, `pure [i, j]`.**
- Expressions which give names to expressions, using the `let` keyword.

---

Now I have the notion of a `do` return result discussed as one of the types of `do` expressions.  If the book wants to discuss `pure [i, j]` vs `[[i, j]]` later, I'm more easily able to digest that because the notion of a `do` return value has already been established.

Reply:
This addition looks good.
If you modify your PR to make this change instead, I'll merge it.

Note that `kind` has another [very specific meaning](https://github.com/purescript-contrib/purescript-book/blob/master/text/chapter3.md#type-constructors-and-kinds) in the language, although I don't think it's much of a problem to use it in the general sense here.

Reply:
Good catch: I'll change `kind` to `category`.  I will also undo all of my other changes, since this is the key change that I think would have helped me.  If you want to keep some of the other changes I made, let me know.  Thanks @milesfrain and @hdgarrood!